tracing-subscriber/src/layer.rs

```diff
@@ -0,0 +1,420 @@
+use tracing_core::{
+    metadata::Metadata,
+    span,
+    subscriber::{Interest, Subscriber},
+    Event,
+};
+
+use std::any::TypeId;
+
+/// A composable handler for `tracing` events.
+///
+/// The [`Subscriber`] trait in `tracing-core` represents the _complete_ set of
+/// functionality required to consume `tracing` instrumentation. This means that
+/// a single `Subscriber` instance is a self-contained implementation of a
+/// complete strategy for collecting traces; but it _also_ means that the
+/// `Subscriber` trait cannot easily be composed with other `Subscriber`s.
+///
+/// In particular, `Subscriber`s are responsible for generating [span IDs] and
+/// assigning them to spans. Since these IDs must uniquely identify a span
+/// within the context of the current trace, this means that there may only be
+/// a single `Subscriber` for a given thread at any point in time &mdash;
+/// otherwise, there would be no authoritative source of span IDs.
+///
+/// On the other hand, the majority of the `Subscriber` trait's functionality
+/// is composable: any number of subscribers may _observe_ events, span entry
+/// and exit, and so on, provided that there is a single authoritative source of
+/// span IDs. The `Layer` trait represents this composable subset of the
+/// `Subscriber` behavior; it can _observe_ events and spans, but does not
+/// assign IDs.
+///
+/// [`Subscriber`]: https://docs.rs/tracing-core/0.1.1/tracing_core//subscriber/trait.Subscriber.html
+/// [span IDs]: https://docs.rs/tracing-core/0.1.1/tracing_core/span/struct.Id.html
+pub trait Layer<S>: 'static {
+    /// Registers a new callsite with this layer, returning whether or not
+    /// the subscriber is interested in being notified about the callsite.
+    ///
+    /// This function is provided with the [`Interest`] returned by the wrapped
+    /// subscriber. The layer may then choose to return that interest, ignore it
+    /// entirely, or combine an `Interest` of its own with the prior `Interest`.
+    ///
+    /// Beyond that, this functions similarly to [`Subscriber::register_callsite`].
+    ///
+    /// By default, this simply returns the `Interest` returned by the wrapped
+    /// subscriber.
+    ///
+    /// [`Interest`]: https://docs.rs/tracing-core/0.1.1/tracing_core/struct.Interest.html
+    /// [`Subscriber::register_callsite`]: https://docs.rs/tracing-core/0.1.1/tracing_core/trait.Subscriber.html#method.register_callsite
+    fn register_callsite(&self, _metadata: &'static Metadata<'static>, prev: Interest) -> Interest {
```

Comment:
I'm not sure I follow why we might need the prev interest? Wouldn't this make it so the order of which you layer the subscribers becomes very important and maybe easy to mess up?

Reply:
> Wouldn't this make it so the order of which you layer the subscribers becomes very important

Yes, but this is _by design_; this is specifically to allow composing multiple filtering layers, or wrapping an existing subscriber with additional filtering. This is the only place in the `Layer` interface where order matters; perhaps that ought to be documented.